datastore/filesystem_datastore.nim

```diff
@@ -0,0 +1,159 @@
+import std/os
+
+import pkg/questionable
+import pkg/questionable/results
+import pkg/upraises
+
+import ./datastore
+
+export datastore
+
+push: {.upraises: [].}
+
+type
+  FileSystemDatastore* = ref object of Datastore
+    root: string
+
+const
+  objExt* = ".obj"
+
+proc new*(
+  T: type FileSystemDatastore,
+  root = "data"): ?!T =
+
+  try:
+    let
+      root = if root.isAbsolute: root
+             else: getCurrentDir() / root
+
+    createDir(root)
+    success T(root: root)
+
+  except IOError as e:
+    failure e
+
+  except OSError as e:
+    failure e
+
+proc root*(self: FileSystemDatastore): string =
+  self.root
+
+proc path*(
+  self: FileSystemDatastore,
+  key: Key): string =
+
+  var
+    segments: seq[string]
+
+  for ns in key:
+    without field =? ns.field:
+      segments.add ns.value
+      continue
+
+    segments.add(field / ns.value)
+
+  # is it problematic that per this logic Key(/a:b) evaluates to the same path
```

Comment:
I say that in the absense of a clear spec arround some of of the behaviours, we should match what other implementations are doing.

Reply:
It seems to vary by implementation as well. While working on a first-pass at `method query` I got an idea re: how to clear this up, will take a stab at it and ask for feedback re: whether it's the behavior we want.